beacon_node/eth2-libp2p/src/rpc/request_response.rs

```diff
@@ -0,0 +1,242 @@
+) -> RPCRequestResponse<TSocket, TData>
+where
+    TSocket: AsyncRead + AsyncWrite,
+    TData: AsRef<[u8]>,
+{
+    RPCRequestResponse {
+        protocol,
+        inner: RPCRequestResponseInner::Write(
+            write_one(socket, data).inner,
+            max_size,
+            short_circuit_return,
+        ),
+    }
+}
+
+/// Future that makes `rpc_request_response` work.
+pub struct RPCRequestResponse<TSocket, TData = Vec<u8>> {
+    protocol: ProtocolId,
+    inner: RPCRequestResponseInner<TSocket, TData>,
+}
+
+enum RPCRequestResponseInner<TSocket, TData> {
+    // We need to write data to the socket.
+    Write(WriteOneInner<TSocket, TData>, usize, Option<RPCResponse>),
+    // We need to read the response code.
+    ReadResponseCode(io::ReadExact<TSocket, io::Window<Vec<u8>>>, usize),
+    // We need to read a final data packet. The second parameter is the response code
+    Read(ReadOne<TSocket>, ResponseCode),
+    // An error happened during the processing.
+    Poisoned,
+}
+
+impl<TSocket, TData> Future for RPCRequestResponse<TSocket, TData>
+where
+    TSocket: AsyncRead + AsyncWrite,
+    TData: AsRef<[u8]>,
+{
+    type Item = RPCResponse;
+    type Error = RPCError;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        loop {
+            match mem::replace(&mut self.inner, RPCRequestResponseInner::Poisoned) {
+                RPCRequestResponseInner::Write(mut inner, max_size, sc_return) => {
+                    match inner.poll().map_err(ReadOneError::Io)? {
+                        Async::Ready(socket) => {
+                            // short-circuit the future if `short_circuit_return` is specified
+                            if let Some(return_val) = sc_return {
+                                return Ok(Async::Ready(return_val));
+                            }
+
+                            // begin reading the 1-byte response code
+                            let mut data_buf = vec![0; 1];
+                            let mut data_buf = io::Window::new(data_buf);
+                            self.inner = RPCRequestResponseInner::ReadResponseCode(
+                                io::read_exact(socket, data_buf),
+                                max_size,
+                            );
+                        }
+                        Async::NotReady => {
+                            self.inner = RPCRequestResponseInner::Write(inner, max_size, sc_return);
+                            return Ok(Async::NotReady);
+                        }
+                    }
+                }
+                RPCRequestResponseInner::ReadResponseCode(mut inner, max_size) => {
+                    match inner.poll()? {
+                        Async::Ready((socket, data)) => {
+                            let resp_code_byte = [0; 1];
+                            // data must be only 1-byte - this cannot panic
+                            resp_code_byte.copy_from_slice(&data.into_inner());
```

Comment:
Out of interest, what restricts `data.into_inner().len()` to `1`?

Reply:
This is a state-machine. In order to get to `ReadResponseCode` state, it is set on line `80`. Notice that `inner` is `io::read_exact(socket, databuf)` where `databuf` is a one byte buffer. Therefore it can only be ready once it's read a single byte. 
This is actually redundant code at the moment, and I will be removing it all.